# Simran Amesar
### Matriculation number - 	100007050

In [8]:
import spacy
from spacy import displacy
from IPython.display import HTML, display

#   Load the English language model
nlp = spacy.load("en_core_web_sm")

text = "I saw the man with a telescope."

doc = nlp(text)

# Print the dependency parse information for each token
for token in doc:
    print(f"{token.text:<12} {token.dep_:<12} {token.head.text}")

html = displacy.render(doc, style="dep", jupyter=False)

# Display the dependency parse visualization in Jupyter Notebook
display(HTML(html))

# Discuss with your partner:
# 1. What is the ROOT?
# 2. What does "with a telescope" modify?
# 3. Does the sentence have more than one meaning?
# 4. Did spaCy choose the interpretation you expected?
# 5. Rewrite the sentence to remove ambiguity.

I            nsubj        saw
saw          ROOT         saw
the          det          man
man          dobj         saw
with         prep         saw
a            det          telescope
telescope    pobj         with
.            punct        saw


### 1. What is the ROOT?
        The root is "saw" - it's the main verb of the sentence.
### 2. What does "with a telescope" modify?
        In spaCy's parse it attaches to saw (as a prep), meaning I used the telescope to see. But it's not fixed — it could also modify 'man'.
### 3. Does the sentence have more than one meaning?
        Yes, either I used a telescope to see the man, or I saw the man who had a telescope.
### 4. Did spaCy choose the interpretation you expected?
        spaCy chose the verb-attachment reading ("saw with a telescope"), which is one valid parse but not the only one.
### 5. Rewrite the sentence to remove ambiguity.
        "I saw the man who had a telescope." OR
        "Using a telescope, I saw the man."

In [9]:
import spacy

nlp = spacy.load("en_core_web_sm")

text = """
Microsoft announced a new AI research initiative
in Berlin on 15 June 2026.

The project involves collaboration with Google,
SAP and SRH University Heidelberg.

Satya Nadella will present the initiative
at a conference in Germany.
"""

doc = nlp(text)

for ent in doc.ents:
    print(
        ent.text,
        ent.label_,
        spacy.explain(ent.label_)
    )

#1. Which entities were detected?
#2. Which labels were assigned?
#3. Which entity types occur most often?
#4. Were any entities missed?

Microsoft ORG Companies, agencies, institutions, etc.
AI ORG Companies, agencies, institutions, etc.
Berlin GPE Countries, cities, states
15 June 2026 DATE Absolute or relative dates or periods
Google ORG Companies, agencies, institutions, etc.
SAP ORG Companies, agencies, institutions, etc.
Satya Nadella PERSON People, including fictional
Germany GPE Countries, cities, states


### 1. Which entities were detected?
       Microsoft, AI, Berlin, 15 June 2026, Google, SAP, Satya Nadella, Germany
### 2. Which labels were assigned?
        ORG, ORG, GPE, DATE, ORG, ORG, PERSON, GPE
### 3. Which entity types occur most often?
        ORG (4 times — Microsoft, AI, Google, SAP)
### 4. Were any entities missed?
        Yes — SRH University Heidelberg was not detected as an entity. 
        Also, AI being labelled ORG is arguably incorrect, it's a concept, not an organisation.

In [10]:

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

documents = [
    "AI is transforming healthcare",
    "Machine learning improves healthcare",
    "Football is a popular sport"
]

vectorizer = TfidfVectorizer()

tfidf_matrix = vectorizer.fit_transform(documents)

print("Vocabulary:")
print(vectorizer.get_feature_names_out())

print("\nTF-IDF Matrix:")
print(tfidf_matrix.toarray())

similarity_matrix = cosine_similarity(tfidf_matrix)

print("\nCosine Similarity Matrix:")
print(similarity_matrix)

print("Doc1 vs Doc2:", round(similarity_matrix[0][1], 3))
print("Doc1 vs Doc3:", round(similarity_matrix[0][2], 3))
print("Doc2 vs Doc3:", round(similarity_matrix[1][2], 3))

# 1. Which pair has the highest similarity?
# 2. Is this result reasonable?
# 3. Why does TF-IDF treat some words as more important?
# 4. Does cosine understand meaning?
# 5. Would it know that "AI" and "machine learning" are related?

Vocabulary:
['ai' 'football' 'healthcare' 'improves' 'is' 'learning' 'machine'
 'popular' 'sport' 'transforming']

TF-IDF Matrix:
[[0.5628291  0.         0.42804604 0.         0.42804604 0.
  0.         0.         0.         0.5628291 ]
 [0.         0.         0.40204024 0.52863461 0.         0.52863461
  0.52863461 0.         0.         0.        ]
 [0.         0.52863461 0.         0.         0.40204024 0.
  0.         0.52863461 0.52863461 0.        ]]

Cosine Similarity Matrix:
[[1.         0.17209173 0.17209173]
 [0.17209173 1.         0.        ]
 [0.17209173 0.         1.        ]]
Doc1 vs Doc2: 0.172
Doc1 vs Doc3: 0.172
Doc2 vs Doc3: 0.0


### 1. Which pair has the highest similarity?
        Doc1 & Doc2 (0.172) and Doc1 & Doc3 (0.172)
### 2. Is this result reasonable?
        Not Completely. Doc1 and Doc2 both mention "healthcare", so sharing a word gives them some similarity. 
        Doc2 and Doc3 sharing no words → 0.0 is expected.
### 3. Why does TF-IDF treat some words as more important?
        Common words like "is" get low weight (high document frequency lowers IDF). 
        Specific words like "transforming" or "healthcare" score higher because they're more distinctive.
### 4. Does cosine understand meaning?
        Cosine only measures overlap of exact token weights.
### 5. Would it know that "AI" and "machine learning" are related?
        No — they are different tokens with no shared surface form, so their vectors have zero overlap despite being close in meaning. 